common/include/common/float_comparison.hpp

```diff
@@ -101,7 +101,72 @@ inline cl::sycl::half abs<cl::sycl::half>(cl::sycl::half value) noexcept {
   return std::abs(static_cast<float>(value));
 }
 
-#endif  // BLAS_DATA_TYPE_HALF
+/**
+ * Custom float/double to uint16_t cast function.
+ */
+template <typename T>
+inline uint16_t cast_to_half(T& val) {
+  static_assert(
+      std::is_scalar<T>::value,
+      "Value to be casted to uint16_t should be either float or double.");
+  // Uint bit representation using 32 (for float) or 64 (for double)
+  using uint_type = typename std::conditional<std::is_same_v<T, float>,
+                                              uint32_t, uint64_t>::type;
+
+  using scalar_t = typename std::remove_const<T>::type;
+
+  int32_t exp;
+  uint_type bin_float, sign, mant;
+
+  // Avoid const qualifier casting cases
+  scalar_t val_temp = const_cast<scalar_t&>(val);
+
+  // Reinterpret to uint for convenient bit manipulation
+  bin_float = *reinterpret_cast<uint32_t*>(&val_temp);
+
+  if constexpr (std::is_same_v<scalar_t, float>) {
+    // Single precision float data case
+    // Extract sign (to half's MSB)
+    sign = (bin_float & 0x80000000) >> 16;
+    // Adjust for half's smaller exponent & bias
+    exp = ((bin_float & 0x7F800000) >> 23) - 127 + 15;
+    // Truncate to 10 MSBits used in half's mantissa
+    mant = (bin_float & 0x007FFFFF) >> 13;
+  } else {
+    // Double precision float data case
+    sign = (bin_float & 0x8000000000000000ULL) >> 48;
+    exp = ((bin_float & 0x7FF0000000000000ULL) >> 52) - 1023 + 15;
+    mant = (bin_float & 0x000FFFFFFFFFFFFFULL) >> 42;
+  }
+
+  // Overflow/Underflow cases
+  if (exp > 31) {
+    // Clamp to max exponent for half (inf)
+    exp = 31;
+    mant = 0;
+  } else if (exp < 0) {
+    // Flush to zero (subnormal not handled)
+    exp = 0;
+    mant = 0;
+  }
+
+  // Reconstruct binary 16-bit output
+  uint16_t out = sign | (exp << 10) | mant;
+
+  return out;
+}
+
+/**
+ * Float/Double to sycl::half cast utility function using the intermediate
+ * cast_to_half(T&).
+ */
+template <typename T>
+inline cl::sycl::half cast_to_sycl_half(T& val) {
+  uint16_t half_bin = cast_to_half(val);
+  cl::sycl::half result = *reinterpret_cast<cl::sycl::half*>(&half_bin);
+  return static_cast<cl::sycl::half>(result);
+}
```

Comment:
Thanks for your patch, @OuadiElfarouki.

```suggestion
template <typename T>
inline cl::sycl::half cast_to_sycl_half(T& val) {
  return static_cast<cl::sycl::half>(val);
}
```
Do we really need a custom function (`cast_to_half`) for casting the halfs?

Reply:
You're right it wasn't necessary. Got it removed!